text/3268-repr-scalable.md

```diff
@@ -0,0 +1,277 @@
+- Feature Name: repr_scalable
+- Start Date: 2022-05-19
+- RFC PR: [rust-lang/rfcs#3268](https://github.com/rust-lang/rfcs/pull/3268)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Expanding the SIMD functionality to allow for runtime determined vector lengths.
+
+# Motivation
+[motivation]: #motivation
+
+Without some support in the compiler it would be impossible to use the
+[ACLE](https://developer.arm.com/architectures/system-architectures/software-standards/acle)
+[SVE](https://developer.arm.com/documentation/102476/latest/) intrinsics from
+Arm.
+
+This RFC will focus on the Arm vector extensions, and will use them for all
+examples. A large amount of what this RFC covers is emitting the vscale
+attribute from LLVM, therefore other scalable vector extensions should work.  In
+an LLVM developer meeting it was mentioned that RISC-V would use what's accepted
+for Arm SVE for their vector extensions.  \[[see slide
+17](https://llvm.org/devmtg/2019-04/slides/TechTalk-Kruppe-Espasa-RISC-V_Vectors_and_LLVM.pdf)\]
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This is an extension to [RFC 1199 SIMD
+Infrastructure](https://rust-lang.github.io/rfcs/1199-simd-infrastructure.html).
+An understanding of that is expected from the reader of this.
+
+## Brief introduction to scalable vectors
+
+A scalable vector type maps to a CPU register where the size of the register
+isn't known during compile time, but it will be a fixed size during
+runtime. Certain properties will be known about this during compile time for
+example, an architecture might specify there is a minimum size to this register
+and that its size must be a multiple of specific amounts, or the maximum
+possible size of the register for instance.
```

Comment:
Is it common to have a maximum size? (I expect not.)

As an earlier stopgap for allowing these types to be put into structs, coroutines (async/generators), and closures: If we had types that were scalable up to some user-defined maximum we could allocate the max size. Perhaps we could work with this, and allow for some fallback behavior if the actual size exceeds the anticipated maximum.

Reply:
All current scalable vector ISAs (that I'm aware of, so mainly Arm SVE/SME and  RISC-V V) have a maximum legal vector length; for SVE and SME that's 2048 bits, although the architecture reserves some space to expand that, and for RVV it's 65536 bits.

Reply:
> and Open-SoC POWER Simple-V

thank you for mentioning the project I helped found and spent 5yr working on (actually named Libre-SoC, formerly Libre-RISCV), unfortunately it's essentially dead now due to people problems, so you can probably ignore it.

Reply:
> actually named Libre-SoC, formerly Libre-RISCV

Thank you for the correction!  I've updated my comment accordingly.  And good to hear definite news about the project from someone involved, even if the news in question is negative.